File: Monsters.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int x, y;
    char move;
};

vector<char> bfs(vector<string>& grid, int n, int m) {
    map<char, pair<int, int>> directions = {{'D', {1, 0}}, {'U', {-1, 0}}, {'L', {0, -1}}, {'R', {0, 1}}};
    queue<pair<int, int>> M, P;
    vector<vector<int>> mt(n, vector<int>(m, -1)), pt(n, vector<int>(m, -1));
    map<pair<int, int>, Node> parent;
    pair<int, int> player;
    bool foundPlayer = false;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == 'M') {
                M.push({i, j});
                mt[i][j] = 0;
            } else if (grid[i][j] == 'A') {
                P.push({i, j});
                pt[i][j] = 0;
                player = {i, j};
                foundPlayer = true;
            }
        }
    }

    if (!foundPlayer) {
        return {};
    }

    if (player.first == 0 || player.second == 0 || player.first == n - 1 || player.second == m - 1) {
        return {'E'};  //
    }

    while (!M.empty()) {
        auto [x, y] = M.front(); M.pop();
        for (auto [_, dir] : directions) {
            int nx = x + dir.first, ny = y + dir.second;
            if (nx >= 0 && nx < n && ny >= 0 && ny < m && grid[nx][ny] == '.' && mt[nx][ny] == -1) {
                mt[nx][ny] = mt[x][y] + 1;
                M.push({nx, ny});
            }
        }
    }

    while (!P.empty()) {
        auto [x, y] = P.front(); P.pop();
        if (x == 0 || y == 0 || x == n - 1 || y == m - 1) {
            vector<char> path;
            while (make_pair(x, y) != player) {
                auto [xp, yp, mo] = parent[{x, y}];
                path.push_back(mo);
                x = xp; y = yp;
            }
            reverse(path.begin(), path.end());
            return path;
        }

        for (auto [move, dir] : directions) {
            int nx = x + dir.first, ny = y + dir.second;
            if (nx >= 0 && nx < n && ny >= 0 && ny < m && grid[nx][ny] == '.' && pt[nx][ny] == -1) {
                if (mt[nx][ny] == -1 || pt[x][y] + 1 < mt[nx][ny]) {
                    pt[nx][ny] = pt[x][y] + 1;
                    P.push({nx, ny});
                    parent[{nx, ny}] = {x, y, move};
                }
            }
        }
    }
    return {};
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<string> grid(n);
    for (int i = 0; i < n; i++) cin >> grid[i];

    vector<char> ans = bfs(grid, n, m);
    if (!ans.empty()) {
        if (ans.size() == 1 && ans[0] == 'E') {
            cout << "YES\n0\n";
        } else {
            cout << "YES\n" << ans.size() << "\n";
            for (char c : ans) cout << c;
            cout << "\n";
        }
    } else {
        cout << "NO\n";
    }
    return 0;
}

File: monsters.py
from collections import deque

def bfs(grid, n, m):
    directions = {'D': (1, 0), 'U': (-1, 0), 'L': (0, -1), 'R': (0, 1)}
    M, P = deque(), deque()
    mt = [[-1] * m for _ in range(n)]
    pt = [[-1] * m for _ in range(n)]
    parent = {}
    player = None

    for i in range(n):
        for j in range(m):
            if grid[i][j] == 'M':
                M.append((i, j))
                mt[i][j] = 0
            elif grid[i][j] == 'A':
                P.append((i, j))
                pt[i][j] = 0
                player = (i, j)

    if player is None:
        return []

    if player[0] == 0 or player[1] == 0 or player[0] == n - 1 or player[1] == m - 1:
        return ['E']

    while M:
        x, y = M.popleft()
        for dx, dy in directions.values():
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == '.' and mt[nx][ny] == -1:
                mt[nx][ny] = mt[x][y] + 1
                M.append((nx, ny))

    # BFS for player
    while P:
        x, y = P.popleft()
        if x == 0 or y == 0 or x == n - 1 or y == m - 1:
            path = []
            while (x, y) != player:
                x, y, move = parent[(x, y)]
                path.append(move)
            path.reverse()
            return path

        for move, (dx, dy) in directions.items():
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == '.' and pt[nx][ny] == -1:
                if mt[nx][ny] == -1 or pt[x][y] + 1 < mt[nx][ny]:
                    pt[nx][ny] = pt[x][y] + 1
                    P.append((nx, ny))
                    parent[(nx, ny)] = (x, y, move)

    return []

def main():
    n, m = map(int, input().split())
    grid = [input().strip() for _ in range(n)]

    ans = bfs(grid, n, m)
    if ans:
        if len(ans) == 1 and ans[0] == 'E':
            print("YES\n0")
        else:
            print("YES")
            print(len(ans))
            print("".join(ans))
    else:
        print("NO")

main()
